File: javascript/sbp/msg.js
/**
 * Copyright (C) 2015 Swift Navigation Inc.
 * Contact: Joshua Gross <josh@swift-nav.com>
 * This source is subject to the license found in the file 'LICENSE' which must
 * be distributed together with this source. All other rights reserved.
 *
 * THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND,
 * EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
 */

var Parser = require('binary-parser').Parser;
var path = require('path');
var streams = require('stream');
var SBP = require('./sbp');

var SBP_PREAMBLE = 0x55;

// Default sender ID. Intended for messages sent from the host to the device.
var SENDER_ID = 0x42;

var crc16tab =  [0x0000,0x1021,0x2042,0x3063,0x4084,0x50a5,0x60c6,0x70e7,
                 0x8108,0x9129,0xa14a,0xb16b,0xc18c,0xd1ad,0xe1ce,0xf1ef,
                 0x1231,0x0210,0x3273,0x2252,0x52b5,0x4294,0x72f7,0x62d6,
                 0x9339,0x8318,0xb37b,0xa35a,0xd3bd,0xc39c,0xf3ff,0xe3de,
                 0x2462,0x3443,0x0420,0x1401,0x64e6,0x74c7,0x44a4,0x5485,
                 0xa56a,0xb54b,0x8528,0x9509,0xe5ee,0xf5cf,0xc5ac,0xd58d,
                 0x3653,0x2672,0x1611,0x0630,0x76d7,0x66f6,0x5695,0x46b4,
                 0xb75b,0xa77a,0x9719,0x8738,0xf7df,0xe7fe,0xd79d,0xc7bc,
                 0x48c4,0x58e5,0x6886,0x78a7,0x0840,0x1861,0x2802,0x3823,
                 0xc9cc,0xd9ed,0xe98e,0xf9af,0x8948,0x9969,0xa90a,0xb92b,
                 0x5af5,0x4ad4,0x7ab7,0x6a96,0x1a71,0x0a50,0x3a33,0x2a12,
                 0xdbfd,0xcbdc,0xfbbf,0xeb9e,0x9b79,0x8b58,0xbb3b,0xab1a,
                 0x6ca6,0x7c87,0x4ce4,0x5cc5,0x2c22,0x3c03,0x0c60,0x1c41,
                 0xedae,0xfd8f,0xcdec,0xddcd,0xad2a,0xbd0b,0x8d68,0x9d49,
                 0x7e97,0x6eb6,0x5ed5,0x4ef4,0x3e13,0x2e32,0x1e51,0x0e70,
                 0xff9f,0xefbe,0xdfdd,0xcffc,0xbf1b,0xaf3a,0x9f59,0x8f78,
                 0x9188,0x81a9,0xb1ca,0xa1eb,0xd10c,0xc12d,0xf14e,0xe16f,
                 0x1080,0x00a1,0x30c2,0x20e3,0x5004,0x4025,0x7046,0x6067,
                 0x83b9,0x9398,0xa3fb,0xb3da,0xc33d,0xd31c,0xe37f,0xf35e,
                 0x02b1,0x1290,0x22f3,0x32d2,0x4235,0x5214,0x6277,0x7256,
                 0xb5ea,0xa5cb,0x95a8,0x8589,0xf56e,0xe54f,0xd52c,0xc50d,
                 0x34e2,0x24c3,0x14a0,0x0481,0x7466,0x6447,0x5424,0x4405,
                 0xa7db,0xb7fa,0x8799,0x97b8,0xe75f,0xf77e,0xc71d,0xd73c,
                 0x26d3,0x36f2,0x0691,0x16b0,0x6657,0x7676,0x4615,0x5634,
                 0xd94c,0xc96d,0xf90e,0xe92f,0x99c8,0x89e9,0xb98a,0xa9ab,
                 0x5844,0x4865,0x7806,0x6827,0x18c0,0x08e1,0x3882,0x28a3,
                 0xcb7d,0xdb5c,0xeb3f,0xfb1e,0x8bf9,0x9bd8,0xabbb,0xbb9a,
                 0x4a75,0x5a54,0x6a37,0x7a16,0x0af1,0x1ad0,0x2ab3,0x3a92,
                 0xfd2e,0xed0f,0xdd6c,0xcd4d,0xbdaa,0xad8b,0x9de8,0x8dc9,
                 0x7c26,0x6c07,0x5c64,0x4c45,0x3ca2,0x2c83,0x1ce0,0x0cc1,
                 0xef1f,0xff3e,0xcf5d,0xdf7c,0xaf9b,0xbfba,0x8fd9,0x9ff8,
                 0x6e17,0x7e36,0x4e55,0x5e74,0x2e93,0x3eb2,0x0ed1,0x1ef0]

// TODO: CRC implementation and verify CRC

var mergeDict = function (dest, src) {
  for (var prop in src) {
    if (src.hasOwnProperty(prop)) {
      dest[prop] = src[prop];
    }
  }
  return dest;
};

function BufferTooShortError (message) {
  this.name = 'BufferTooShortError';
  this.message = message;
  this.stack = (new Error()).stack;
}
BufferTooShortError.prototype = Object.create(Error.prototype);
BufferTooShortError.prototype.constructor = BufferTooShortError;

function BufferCorruptError (message) {
  this.name = 'BufferCorruptError';
  this.message = message;
  this.stack = (new Error()).stack;
}
BufferCorruptError.prototype = Object.create(Error.prototype);
BufferCorruptError.prototype.constructor = BufferCorruptError;

var sbpImports = {
  acquisition: require('./acquisition.js'),
  bootload: require('./bootload.js'),
  ext_events: require('./ext_events.js'),
  file_io: require('./file_io.js'),
  flash: require('./flash.js'),
  logging: require('./logging.js'),
  navigation: require('./navigation.js'),
  observation: require('./observation.js'),
  piksi: require('./piksi.js'),
  settings: require('./settings.js'),
  system: require('./system.js'),
  tracking: require('./tracking.js')
};

var sbpIdTable = Object.keys(sbpImports).reduce(function (prev, key) {
  var curr = sbpImports[key];
  var numericKeysDict = {};
  Object.keys(curr).map(function (key) {
    if (parseInt(key) == key) {
      numericKeysDict[key] = curr[key];
    }
  });
  return mergeDict(prev, numericKeysDict);
}, {});

var sbpMessageTypesTable = Object.keys(sbpImports).reduce(function (prev, key) {
  var curr = sbpImports[key];
  var nonNumericKeysDict = {};
  Object.keys(curr).map(function (key) {
    if (parseInt(key) != key) {
      var messageType = curr[key].prototype.messageType || key;
      nonNumericKeysDict[messageType] = curr[key];
    }
  });
  return mergeDict(prev, nonNumericKeysDict);
}, {});

var parser = new Parser()
  .endianess('little')
  .uint8('preamble')
  .uint16('msg_type')
  .uint16('sender')
  .uint8('length')
  .buffer('payload', { length: 'length' })
  .uint16('crc');

/**
 * CRC16 implementation according to CCITT standards.
 * Based on msg.py implementation.
 */
function crc16 (buf, crc) {
  crc = crc || 0;

  for (var i = 0; i < buf.length; i++) {
    var ch = buf[i];
    crc = ((crc<<8)&0xFFFF) ^ (crc16tab[((crc>>8)&0xFF) ^ (ch&0xFF)]);
    crc &= 0xFFFF;
  }

  return crc;
}

module.exports = {
  preambleByte: SBP_PREAMBLE,
  crc16: crc16,

  sbpIdTable: sbpIdTable,
  sbpMessageTypesTable: sbpMessageTypesTable,

  decode: function decode (msg) {
    var sbp = parser.parse(msg);
    var msgTypeDecoder = sbpIdTable[sbp['msg_type']];

    if (typeof msgTypeDecoder === 'undefined') {
      console.log("Unknown message type: ", sbp['msg_type']);
      return new SBP(sbp);
    }

    return new msgTypeDecoder(sbp);
  },
  /**
   * Look for the beginning of a framed message in a stream, and parse it until the end. Verify
   *  that the CRC matches the message. If the framed message is good, invoke callback. Otherwise throw message away.
   *  Repeats until the stream has ended.
   *
   * Callback will be called with valid framed messages. Will not currently be called with errors.
   *
   * This corresponds to the logic in `framer.py`.
   *
   * @param stream: A Readable stream of bytes.
   * @param callback: a callback function invoked when a framed message is found and decoded in the stream.
   * @returns [parsed SBP object, Buffer]
   */
  dispatch: function dispatch (stream, callback) {
    var offset = 0;
    var streamBuffer = new Buffer(0);

    var getFramedMessage = function () {
      var headerBuf, payloadBuf;
      var preamble, msgType, sender, length, crc;
      var payloadCrc;

      // Find preamble byte
      var preamblePos;
      for (preamblePos = 0; preamblePos < streamBuffer.length; preamblePos++) {
        preamble = streamBuffer.readUInt8(preamblePos);
        if (preamble == SBP_PREAMBLE) {
          break;
        }
      }
      if (preamble != SBP_PREAMBLE) {
        throw new BufferTooShortError();
      }

      // Get msg_type, sender, length next
      if (preamblePos+6 > streamBuffer.length) {
        throw new BufferTooShortError();
      }
      headerBuf = streamBuffer.slice(preamblePos+1, preamblePos+6);
      msgType = streamBuffer.readUInt16LE(preamblePos+1);
      sender = streamBuffer.readUInt16LE(preamblePos+3);
      length = streamBuffer.readUInt8(preamblePos+5);

      // Get full payload
      // First, check payload length + CRC
      if (preamblePos+8+length > streamBuffer.length) {
        throw new BufferTooShortError();
      }
      payloadBuf = streamBuffer.slice(preamblePos+6, preamblePos+6+length);
      payloadCrc = crc16(payloadBuf, crc16(headerBuf));

      // Finally, get and verify CRC
      crc = streamBuffer.readUInt16LE(preamblePos+6+length);

      // Pull out full buffer
      var fullBuffer = streamBuffer.slice(preamblePos, preamblePos+6+length+2);

      // If the CRC looks correct, decode the full payload
      if (crc === payloadCrc) {
        // remove this chunk from the stream buffer
        streamBuffer = streamBuffer.slice(preamblePos+6+length+2);

        return [module.exports.decode(fullBuffer), fullBuffer];
      } else {
        // remove bad preamble from the stream buffer
        streamBuffer = streamBuffer.slice(preamblePos+1);

        throw new BufferCorruptError();
      }
    };

    var processData = function (data) {
      stream.pause();
      try {
        streamBuffer = Buffer.concat([streamBuffer, data]);
        if (streamBuffer.length < 2) {
          return;
        }
        var pair = getFramedMessage();
        var framedMessage = pair[0];
        var fullBuffer = pair[1];

        // If there is data left to process after a successful parse, process again
        if (streamBuffer.length > 0) {
          setTimeout(function () {
            processData(new Buffer(0));
          }, 0);
        }

        callback(null, framedMessage, fullBuffer);
      } catch (e) {
        // If the buffer was corrupt but there's more in the stream, try again immediately
        if (e instanceof BufferCorruptError && streamBuffer.length > 0) {
          setTimeout(function () {
            processData(new Buffer(0));
          }, 0);
        }
      } finally {
        offset = 0;
        stream.resume();
      }
    };

    stream.on('data', processData);
  }
};

File: python/sbp/msg.py
#!/usr/bin/env python
# Copyright (C) 2011-2021 Swift Navigation Inc.
# Contact: https://support.swiftnav.com
#
# This source is subject to the license found in the file 'LICENSE' which must
# be be distributed together with this source. All other rights reserved.
#
# THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND,
# EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED
# WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.

import binascii
import copy
import json
import struct
try:
  import pybase64 as base64
except ImportError:
  import base64

from sbp.constants import SENDER_ID as _SENDER_ID
from sbp.constants import SBP_PREAMBLE as _SBP_PREAMBLE


SENDER_ID = _SENDER_ID
SBP_PREAMBLE = _SBP_PREAMBLE

_HEADER_FMT = '<BHHB'
_HEADER_LEN = struct.calcsize(_HEADER_FMT)
_HEADER_PARSER = struct.Struct(_HEADER_FMT)

_CRC_FMT = '<H'
_CRC_LEN = struct.calcsize(_CRC_FMT)
_CRC_PARSER = struct.Struct(_CRC_FMT)


def try_import_jit():
  from warnings import warn

  warn("sbp.jit has been removed", UserWarning, stacklevel=1)
  return None


class UnpackError(ValueError):
  pass


def crc16(d, s=0):
  return binascii.crc_hqx(d, s)


class _StreamPayload(object):

  def __init__(self):
    self.length = None
    self.buffer = None
    self.offset = None

  def reset(self, buffer_, offset):
    self.length = 0
    self.buffer = buffer_
    self.offset = offset

  def write(self, data):
    length = len(data)
    self.buffer[self.offset:self.offset + length] = bytearray(data)
    self.length += length
    self.offset += length
    return length


class SBP(object):
  """Swift Binary Protocol container.

  """

  __slots__ = ['preamble',
               'msg_type',
               'sender',
               'length',
               'payload',
               'crc',
               'stream_payload',
               'parser']

  def __init__(self, msg_type=None, sender=SENDER_ID,
               length=None, payload=None, crc=None):
    self.preamble = SBP_PREAMBLE
    self.msg_type = msg_type
    self.sender = sender
    self.length = length
    self.payload = payload
    self.crc = crc
    self.stream_payload = _StreamPayload()
    self.parser = None

  def __eq__(self, other):
    try:
      if self is other:
        return True
      elif not isinstance(self, type(other)):
        return False
      elif self.__slots__ != other.__slots__:
        return False
      else:
        return all(getattr(self, s) == getattr(other, s) for s in self.__slots__)
    except AttributeError:
      return False

  def __update(self):
    raise NotImplementedError("Internal update used by children.")

  def _copy_payload(self, buf, offset, payload):
    length = len(payload)
    end_offset = offset + length
    buf[offset:end_offset] = bytearray(payload)
    return length

  def _get_framed(self, buf, offset, insert_payload):
    """Returns the framed message and updates the CRC.

    """
    header_offset = offset + _HEADER_LEN
    self.length = insert_payload(buf, header_offset, self.payload)
    _HEADER_PARSER.pack_into(buf,
                             offset,
                             self.preamble,
                             self.msg_type,
                             self.sender,
                             self.length)
    crc_offset = header_offset + self.length
    preamble_bytes = 1
    crc_over_len = _HEADER_LEN + self.length - preamble_bytes
    self.crc = crc16(buf[1+offset:1+offset+crc_over_len])
    _CRC_PARSER.pack_into(buf, crc_offset, self.crc)
    length = preamble_bytes + crc_over_len + _CRC_LEN
    return length

  def pack(self):
    """Pack to framed binary message."""
    buf = bytearray(512)
    packed_len = self._get_framed(buf, 0, self._copy_payload)
    return bytes(buf[:packed_len])

  def pack_into(self, buf, offset, write_payload):
    """Pack to framed binary message."""
    return self._get_framed(buf, offset, write_payload)

  @staticmethod
  def calc_crc(msg_type, sender, payload):
    msgbuf = bytearray(256)
    _HEADER_PARSER.pack_into(msgbuf, 0, SBP_PREAMBLE, msg_type, sender, len(payload))
    msgbuf[_HEADER_LEN:_HEADER_LEN+len(payload)] = payload
    return crc16(msgbuf[1:_HEADER_LEN+len(payload)])

  @staticmethod
  def unpack(d):
    """Unpack and return a framed binary message."""
    try:
      preamble, msg_type, sender, length = _HEADER_PARSER.unpack(d[:_HEADER_LEN])
    except struct.error:
      raise UnpackError("invalid header")
    assert preamble == SBP_PREAMBLE, "Invalid preamble 0x%x." % preamble
    payload = d[_HEADER_LEN:_HEADER_LEN+length]
    try:
      (crc,) = _CRC_PARSER.unpack(d[_HEADER_LEN+length:_HEADER_LEN+length+_CRC_LEN])
    except struct.error:
      raise UnpackError("invalid CRC")
    computed_crc = SBP.calc_crc(msg_type, sender, payload)
    if crc != computed_crc:
      exc = UnpackError("CRC error: computed {}, parsed {}".format(computed_crc, crc))
      exc.malformed_msg = SBP(msg_type, sender, length, payload, crc)
      raise exc
    return SBP(msg_type, sender, length, payload, crc)

  def copy(self):
    return copy.deepcopy(self)

  def __repr__(self):
    p = (self.preamble, self.msg_type, self.sender, self.length,
         repr(self.payload), self.crc)
    fmt = "<SBP (preamble=0x%X, msg_type=0x%X, sender=%s, length=%d, payload=%s, crc=0x%X)>"
    return fmt % p

  def to_binary(self):
    ret = _HEADER_PARSER.pack(SBP_PREAMBLE, self.msg_type, self.sender, len(self.payload))
    ret += self.payload
    crc = crc16(ret[1:])
    ret += _CRC_PARSER.pack(crc)
    return ret

  def to_json(self, sort_keys=False):
    """Produce a JSON-encoded SBP message."""
    d = self.to_json_dict()
    return json.dumps(d, sort_keys=sort_keys)

  @staticmethod
  def from_json(s):
    """Given a JSON-encoded message, build an object."""
    d = json.loads(s)
    sbp = SBP.from_json_dict(d)
    return sbp

  @staticmethod
  def from_json_dict(d):
    sbp = SBP()
    sbp.msg_type = d.pop('msg_type')
    sbp.sender = d.pop('sender')
    sbp.length = d.pop('length')
    sbp.payload = base64.standard_b64decode(d.pop('payload'))
    sbp.crc = d.pop('crc')
    return sbp

  def to_json_dict(self):
    d = {'preamble': self.preamble,
         'msg_type': self.msg_type,
         'sender': self.sender,
         'length': self.length,
         'payload': base64.standard_b64encode(self.payload).decode('ascii'),
         'crc': self.crc}
    return d

  def _build_payload(self, buf, offset, payload):
    self.stream_payload.reset(buf, offset)
    self.parser.build_stream(payload, self.stream_payload)
    return self.stream_payload.length

  def _write_payload(self, buf, offset, payload):
    self.stream_payload.reset(buf, offset)
    self.stream_payload.write(payload)
    return self.stream_payload.length

  def into_buffer(self, buf, offset):
    if self.payload:
      return self.pack_into(buf, offset, self._write_payload)
    else:
      def _empty_payload(_buf, _offset, _payload):
        return 0
      return self.pack_into(buf, offset, _empty_payload)
